Shell/Formatter.cpp

```diff
@@ -0,0 +1,648 @@
+/*
+ * Copyright (c) 2020, the SerenityOS developers.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "Formatter.h"
+#include "AST.h"
+#include "Parser.h"
+#include <AK/TemporaryChange.h>
+
+String Formatter::format()
+{
+    auto node = Parser(m_source).parse();
+    if (m_cursor >= 0)
+        m_output_cursor = m_cursor;
+
+    if (!node)
+        return String();
+
+    if (node->is_syntax_error())
+        return m_source;
+
+    if (m_cursor >= 0) {
+        auto hit_test = node->hit_test_position(m_cursor);
+        if (hit_test.matching_node)
+            m_hit_node = hit_test.matching_node.ptr();
+        else
+            m_hit_node = nullptr;
+    }
+
+    m_should_commit = true;
+
+    m_parent_node = node;
+
+    node->visit(*this);
+
+    m_should_commit = false;
+
+    auto string = m_builder.string_view();
+
+    if (!string.ends_with(" "))
+        m_builder.append(m_trivia);
+
+    return m_builder.to_string();
+}
+
+void Formatter::test_and_update_output_cursor(const AST::Node* node)
+{
+    if (!node)
+        return;
+
+    if (node != m_hit_node)
+        return;
+
+    m_output_cursor = current_builder().length() + m_cursor - node->position().start_offset;
+}
+
+void Formatter::insert_separator()
+{
+    current_builder().append('\n');
+    insert_indent();
+}
+
+void Formatter::insert_indent()
+{
+    for (size_t i = 0; i < m_current_indent; ++i)
+        current_builder().append("  ");
```

Comment:
I believe there are two spaces there. Do you mean it?

Reply:
Yes, I decided to initially indent with two spaces, later we can make that an option.